### PR TITLE
[Serverless - WIP] Remove STS call ans refactor tags

### PR DESCRIPTION
**Pull request by maxday**
[work in progress]

### What does this PR do?

 - Remove the STS call which was used to compute tags before the first invocation. Now tags are computed during the first invocation thanks to the extension API payload field  `InvokedFunctionArn`
 - Refactor tag management

Need to verify :
🟠    behaviour on us-gov
✅    behaviour with VPC
✅    thread safety of tag computing during the first invocations

### Motivation

Increase performance of extension startup + code cleaning

### Describe how to test your changes

Need to document + summarize performance improvements

### CODE REVIEW

File: pkg/serverless/aws/arn.go [hunk 1]
```diff
@@ -172,103 +141,3 @@ func RestoreCurrentStateFromFile() error {
 	SetRequestID(restoredState.CurrentReqID)
 	return nil
 }
-
-// FetchFunctionARNFromEnv reconstructs the function arn from what's available
-// in the environment.
-func FetchFunctionARNFromEnv(accountID string) (string, error) {
```
Comment (maxday) [inline pkg/serverless/aws/arn.go:178]: need to double check the behaviour in us-gov but my understanding is that we can get rid of this logic since `InvokedFunctionArn` will be already well formatted (and prefixed with `aws-us-gov`)

File: pkg/serverless/protocol.go [hunk 2]
```diff
@@ -273,6 +288,24 @@ func (d *Daemon) WaitUntilClientReady(timeout time.Duration) bool {
 	return d.clientLibReady
 }
 
+// ComputeGlobalTags extracts tags from the ARN, merge them use any user-predefined tags and set them to traces, logs and metrics
+func (d *Daemon) ComputeGlobalTags(arn string, tags []string) {
```
Comment (DarcyRaynerDD) [inline pkg/serverless/protocol.go:292]: What's the casing of functionArn here? Usually, we make sure the tags are lower case, including functionArn.

File: pkg/serverless/protocol.go [hunk 3]
```diff
@@ -207,10 +218,12 @@ func StartDaemon(stopTraceAgent context.CancelFunc) *Daemon {
 		mux:              mux,
 		ReadyWg:          &sync.WaitGroup{},
 		InvcWg:           &sync.WaitGroup{},
+		extraTagsSetWg:   &sync.WaitGroup{},
```
Comment (DarcyRaynerDD) [inline pkg/serverless/protocol.go:221]: Would it make sense to  reuse the ReadyWg instead? The /hello route blocks until the ReadyWg completes, so it could be useful to make sure the handler code doesn't start until we are ready.

**Review by DarcyRaynerDD — CHANGES_REQUESTED** [review]
Left some comments

File: pkg/serverless/protocol.go (hunk 3 above)
Comment (maxday) [reply pkg/serverless/protocol.go]: Totally ! 💯

File: pkg/serverless/protocol.go (hunk 2 above)
Comment (maxday) [reply pkg/serverless/protocol.go:285]: Yes it's lower cased downstream here : https://github.com/DataDog/datadog-agent/pull/8296/files#diff-bf4b276d7656d4b9786de3d61381eaaa8270a2d65b6a49f6f466a473b9dbd7b2R85

Comment (DarcyRaynerDD) [reply pkg/serverless/protocol.go]: Could we add a test case with an uppercased functionName to be safe?

Comment (maxday) [reply pkg/serverless/protocol.go]: ✅

File: pkg/serverless/tags_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	assert.Equal(t, "lambda", tagMap["_dd.origin"])
+	assert.Equal(t, "1", tagMap["_dd.compute_stats"])
+	assert.Equal(t, "arn:aws:lambda:us-east-1:123456789012:function:my-function", tagMap["function_arn"])
+	assert.Equal(t, "us-east-1", tagMap["region"])
+	assert.Equal(t, "123456789012", tagMap["aws_account"])
+	assert.Equal(t, "my-function", tagMap["functionname"])
+	assert.Equal(t, "my-function", tagMap["resource"])
+}
+
+func TestBuildTagMapFromArnCompleteWithUperCase(t *testing.T) {
```
Comment (DarcyRaynerDD) [inline pkg/serverless/tags_test.go:84]:
```suggestion
func TestBuildTagMapFromArnCompleteWithUpperCase(t *testing.T) {
```

**Review by DarcyRaynerDD — APPROVED** [review]
LGTM

File: pkg/serverless/tags_test.go (hunk 4 above)
Comment (maxday) [reply pkg/serverless/tags_test.go]: ✅

File: pkg/dogstatsd/server.go [hunk 5]
```diff
@@ -833,3 +833,8 @@ func FormatDebugStats(stats []byte) (string, error) {
 
 	return buf.String(), nil
 }
+
+// SetExtraTags sets extra tags to the server
```
Comment (nhinsch) [inline pkg/dogstatsd/server.go:837]: I would rephrase this comment -- what are the extra tags set on exactly? I wouldn't say they are set "to the server"

File: pkg/serverless/enhanced_metrics.go [hunk 6]
```diff
@@ -119,8 +119,7 @@ func sendTimeoutEnhancedMetric(tags []string, metricsChan chan []metrics.MetricS
 }
 
 // getTagsForEnhancedMetrics returns the tags that should be included with enhanced metrics
-func getTagsForEnhancedMetrics() []string {
-	tags := aws.GetARNTags()
+func getTagsForEnhancedMetrics(tags []string) []string {
```
Comment (nhinsch) [inline pkg/serverless/enhanced_metrics.go:122]: I would consider renaming this function -- it looks like it now takes a list of tags and adds a cold start tag. Just from the function name, it's not obvious what tags I need to pass in. Maybe it should just be called `addColdStartTag`?

File: pkg/serverless/protocol.go [hunk 7]
```diff
@@ -273,6 +281,28 @@ func (d *Daemon) WaitUntilClientReady(timeout time.Duration) bool {
 	return d.clientLibReady
 }
 
+// ComputeGlobalTags extracts tags from the ARN, merge them use any user-predefined tags and set them to traces, logs and metrics
```
Comment (nhinsch) [inline pkg/serverless/protocol.go:284]:
```suggestion
// ComputeGlobalTags extracts tags from the ARN, merges them with any user-defined tags and adds them to traces, logs and metrics
```

File: pkg/trace/agent/agent.go [hunk 8]
```diff
@@ -439,3 +439,8 @@ func newEventProcessor(conf *config.AgentConfig) *event.Processor {
 
 	return event.NewProcessor(extractors, conf.MaxEPS)
 }
+
+// SetGlobalTags sets global tags to the current configuration
```
Comment (nhinsch) [inline pkg/trace/agent/agent.go:443]: Does this really set global tags to the current configuration? It looks like it sets them to whatever is passed into this function.

File: pkg/serverless/tags.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func setIfNotEmpty(tagMap map[string]string, key string, value string) map[string]string {
+	if key != "" {
+		tagMap[key] = strings.ToLower(value)
+	}
+	return tagMap
+}
+
+func blackListTagForMetricsAndLogs() []string {
```
Comment (nhinsch) [inline pkg/serverless/tags.go:91]: What do these "blacklist" functions do?

File: cmd/serverless/main.go [hunk 10]
```diff
@@ -334,10 +316,11 @@ func runAgent(stopCh chan struct{}) (daemon *serverless.Daemon, err error) {
 	aggregatorInstance := aggregator.InitAggregator(serializer, nil, "serverless")
 	metricsChan := aggregatorInstance.GetBufferedMetricsWithTsChannel()
 
+	config.Datadog.Set("dogstatsd_packet_buffer_flush_timeout", 1*time.Millisecond)
```
[the remark this answers is not in the document]
Comment (maxday) [reply cmd/serverless/main.go:319]: good catch I've already merged the other one, I'm going to remove it from this PR 👍

File: pkg/serverless/protocol.go (hunk 7 above)
Comment (maxday) [reply pkg/serverless/protocol.go]: ✅

File: pkg/trace/agent/agent.go (hunk 8 above)
Comment (maxday) [reply pkg/trace/agent/agent.go]: ✅

File: pkg/serverless/enhanced_metrics.go (hunk 6 above)
Comment (maxday) [reply pkg/serverless/enhanced_metrics.go]: ✅

File: pkg/serverless/tags.go (hunk 9 above)
Comment (maxday) [reply pkg/serverless/tags.go:91]: Since we now parse one single time the ARN to compute tags from traces metrics and log, those methods are here to prevent attaching some useless tags to logs. For instance `_dd.origin` and `_dd.compute_stats` are useful for traces but not for logs and metrics

Comment (nhinsch) [reply pkg/serverless/tags.go:91]: Aren't these basically just constants though?

Comment (maxday) [reply pkg/serverless/tags.go:91]: Yes these are constants but they are useful for traces but totally useless for both tags and metrics (in the current implementation, we do not attend those tags to metrics or logs)

Comment (maxday) [reply pkg/serverless/tags.go]: -> changed to static arrays 
✅
